pkg/v1/stream/layer.go

```diff
@@ -159,13 +159,17 @@ func newCompressedReader(l *Layer) (*compressedReader, error) {
 	}
 
 	cr := &compressedReader{
-		closer: newMultiCloser(zw, l.blob),
-		pr:     pr,
-		bw:     bw,
-		h:      h,
-		zh:     zh,
-		count:  count,
-		l:      l,
+		// NOTE: Order matters! If zw is closed first, then it will panic,
+		// because io.Copy in the goroutine below will still be copying the
+		// contents of l.blob into it.
```

Comment:
I found this out when calling `Close` on a `stream.Layer` that had not been fully read. I can write a test, but it's getting a bit out of scope for this PR.

And this stops a panic from occurring. It was still confusing to see a `stream.ErrNotConsumed` after calling `Close()` even though it technically makes sense.

Reply:
I'm really surprised nobody has caught this yet. Where do we panic with the other order?

Reply:
Try this. Revert the change:

```diff
diff --git a/pkg/v1/stream/layer.go b/pkg/v1/stream/layer.go
index 9f42f630..3a5993f7 100644
--- a/pkg/v1/stream/layer.go
+++ b/pkg/v1/stream/layer.go
@@ -162,7 +162,8 @@ func newCompressedReader(l *Layer) (*compressedReader, error) {
 		// NOTE: Order matters! If zw is closed first, then it will panic,
 		// because io.Copy in the goroutine below will still be copying the
 		// contents of l.blob into it.
-		closer: newMultiCloser(l.blob, zw),
+		closer: newMultiCloser(zw, l.blob),
+		// closer: newMultiCloser(l.blob, zw),
 
 		pr:    pr,
 		bw:    bw,
```

Add this test:

```go
func TestCloseBeforeConsume(t *testing.T) {
	// Write small tar to pipe
	pr, pw := io.Pipe()
	tw := tar.NewWriter(pw)
	go func() {
		pw.CloseWithError(func() error {
			body := "test file"
			if err := tw.WriteHeader(&tar.Header{
				Name:     "test.txt",
				Mode:     0600,
				Size:     int64(len(body)),
				Typeflag: tar.TypeReg,
			}); err != nil {
				return err
			}
			if _, err := tw.Write([]byte(body)); err != nil {
				return err
			}
			return tw.Close()
		}())
	}()

	// Create stream layer from tar pipe
	l := NewLayer(pr)
	rc, err := l.Compressed()
	if err != nil {
		t.Fatalf("Compressed: %v", err)
	}

	// Close stream layer before consuming
	if err := rc.Close(); err != nil {
		t.Errorf("Close: %v", err)
	}
}
```

If you run it, you'll see that it panics maybe 15-30% of the time. Below, it panicked on the third run.

```
$ go test -count=20 -run CloseBeforeConsume ./pkg/v1/stream/
--- FAIL: TestCloseBeforeConsume (0.00s)
    layer_test.go:257: Close: io: read/write on closed pipe
--- FAIL: TestCloseBeforeConsume (0.00s)
    layer_test.go:257: Close: io: read/write on closed pipe
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x4ebeb4]

goroutine 14 [running]:
compress/flate.(*Writer).Write(...)
        /home/bkrieger/.gimme/versions/go1.17.2.linux.amd64/src/compress/flate/deflate.go:712
compress/gzip.(*Writer).Write(0xc0000b6840, {0xc0002c0000, 0x200, 0x8000})
        /home/bkrieger/.gimme/versions/go1.17.2.linux.amd64/src/compress/gzip/gzip.go:196 +0x334
io.(*multiWriter).Write(0x40b54d, {0xc0002c0000, 0x200, 0x8000})
        /home/bkrieger/.gimme/versions/go1.17.2.linux.amd64/src/io/multi.go:60 +0x86
io.copyBuffer({0x5d1b00, 0xc0002be000}, {0x5d1a60, 0xc0000100c0}, {0x0, 0x0, 0x0})
        /home/bkrieger/.gimme/versions/go1.17.2.linux.amd64/src/io/io.go:425 +0x204
io.Copy(...)
        /home/bkrieger/.gimme/versions/go1.17.2.linux.amd64/src/io/io.go:382
github.com/google/go-containerregistry/pkg/v1/stream.newCompressedReader.func1()
        /home/bkrieger/go/src/github.com/google/go-containerregistry/pkg/v1/stream/layer.go:176 +0xe7
created by github.com/google/go-containerregistry/pkg/v1/stream.newCompressedReader
        /home/bkrieger/go/src/github.com/google/go-containerregistry/pkg/v1/stream/layer.go:175 +0x575
FAIL    github.com/google/go-containerregistry/pkg/v1/stream    0.007s
FAIL
```

Running the test with the race detector helps explain a bit further.

Reply:
Interesting... this looks like it might be a race in the gzip package where concurrently calling Write and Close isn't safe. Nice find.

Reply:
https://go.dev/play/p/H9WwjIFCs9C

Reply:
Looks like this is working as intended but I disagree: https://github.com/golang/go/issues/18883

IMO Write or Close should return an error and not panic here. I'll send a PR for compress/gzip.

Reply:
Something I was wondering as I explored the code: why does `Compressed()` and similar functions return an `io.ReadCloser` rather than an `io.Reader`? And why do functions like `layout.WriteBlob` take an `io.ReadCloser`? It doesn't seem like the `Close()` method is ever or should ever be called by a user of this library.

Reply:
Calling `Close()` allows us to clean up e.g. the goroutines that are gzipping stuff and writing into an `io.PipeWriter` or an open file. Otherwise we'd just leak resources sometimes.

Reply:
I see. I guess that's true for tarball and remote layers. It's just stream layers that close themselves automatically once fully read. And `layout.WriteBlob` probably should have just taken a reader since cleanup duties are generally deferred to the opener, but now the API is set.